File: tests/rosetta/transpiler/php/brazilian-numbers.php
<?php
ini_set('memory_limit', '-1');
function _str($x) {
    if (is_array($x)) {
        $isList = array_keys($x) === range(0, count($x) - 1);
        if ($isList) {
            $parts = [];
            foreach ($x as $v) { $parts[] = _str($v); }
            return '[' . implode(' ', $parts) . ']';
        }
        $parts = [];
        foreach ($x as $k => $v) { $parts[] = _str($k) . ':' . _str($v); }
        return 'map[' . implode(' ', $parts) . ']';
    }
    if (is_bool($x)) return $x ? 'true' : 'false';
    if ($x === null) return 'null';
    return strval($x);
}
function sameDigits($n, $b) {
  global $isBrazilian, $isPrime, $main;
  $f = $n % $b;
  $n = intval((intdiv($n, $b)));
  while ($n > 0) {
  if ($n % $b != $f) {
  return false;
}
  $n = intval((intdiv($n, $b)));
};
  return true;
}
function isBrazilian($n) {
  global $sameDigits, $isPrime, $main;
  if ($n < 7) {
  return false;
}
  if ($n % 2 == 0 && $n >= 8) {
  return true;
}
  $b = 2;
  while ($b < $n - 1) {
  if (sameDigits($n, $b)) {
  return true;
}
  $b = $b + 1;
};
  return false;
}
function isPrime($n) {
  global $sameDigits, $isBrazilian, $main;
  if ($n < 2) {
  return false;
}
  if ($n % 2 == 0) {
  return $n == 2;
}
  if ($n % 3 == 0) {
  return $n == 3;
}
  $d = 5;
  while ($d * $d <= $n) {
  if ($n % $d == 0) {
  return false;
}
  $d = $d + 2;
  if ($n % $d == 0) {
  return false;
}
  $d = $d + 4;
};
  return true;
}
function main() {
  global $sameDigits, $isBrazilian, $isPrime;
  $kinds = [' ', ' odd ', ' prime '];
  foreach ($kinds as $kind) {
  echo rtrim('First 20' . $kind . 'Brazilian numbers:'), PHP_EOL;
  $c = 0;
  $n = 7;
  while (true) {
  if (isBrazilian($n)) {
  echo rtrim(_str($n) . ' '), PHP_EOL;
  $c = $c + 1;
  if ($c == 20) {
  echo rtrim('
'), PHP_EOL;
  break;
};
}
  if ($kind == ' ') {
  $n = $n + 1;
} else {
  if ($kind == ' odd ') {
  $n = $n + 2;
} else {
  while (true) {
  $n = $n + 2;
  if (isPrime($n)) {
  break;
}
};
};
}
};
};
  $n = 7;
  $c = 0;
  while ($c < 100000) {
  if (isBrazilian($n)) {
  $c = $c + 1;
}
  $n = $n + 1;
};
  echo rtrim('The 100,000th Brazilian number: ' . _str($n - 1)), PHP_EOL;
}
main();

File: tests/rosetta/transpiler/FS/brazilian-numbers.fs
// Generated 2025-07-26 04:38 +0700

exception Break
exception Continue

exception Return

let rec sameDigits (n: int) (b: int) =
    let mutable __ret : bool = Unchecked.defaultof<bool>
    let mutable n = n
    let mutable b = b
    try
        let mutable f: int = ((n % b + b) % b)
        n <- int (n / b)
        while n > 0 do
            if (((n % b + b) % b)) <> f then
                __ret <- false
                raise Return
            n <- int (n / b)
        __ret <- true
        raise Return
        __ret
    with
        | Return -> __ret
and isBrazilian (n: int) =
    let mutable __ret : bool = Unchecked.defaultof<bool>
    let mutable n = n
    try
        if n < 7 then
            __ret <- false
            raise Return
        if ((((n % 2 + 2) % 2)) = 0) && (n >= 8) then
            __ret <- true
            raise Return
        let mutable b: int = 2
        while b < (n - 1) do
            if unbox<bool> (sameDigits n b) then
                __ret <- true
                raise Return
            b <- b + 1
        __ret <- false
        raise Return
        __ret
    with
        | Return -> __ret
and isPrime (n: int) =
    let mutable __ret : bool = Unchecked.defaultof<bool>
    let mutable n = n
    try
        if n < 2 then
            __ret <- false
            raise Return
        if (((n % 2 + 2) % 2)) = 0 then
            __ret <- n = 2
            raise Return
        if (((n % 3 + 3) % 3)) = 0 then
            __ret <- n = 3
            raise Return
        let mutable d: int = 5
        while (d * d) <= n do
            if (((n % d + d) % d)) = 0 then
                __ret <- false
                raise Return
            d <- d + 2
            if (((n % d + d) % d)) = 0 then
                __ret <- false
                raise Return
            d <- d + 4
        __ret <- true
        raise Return
        __ret
    with
        | Return -> __ret
and main () =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    try
        let mutable kinds: string array = [|" "; " odd "; " prime "|]
        for kind in kinds do
            try
                printfn "%s" (("First 20" + (unbox<string> kind)) + "Brazilian numbers:")
                let mutable c: int = 0
                let mutable n: int = 7
                try
                    while true do
                        if unbox<bool> (isBrazilian n) then
                            printfn "%s" ((string n) + " ")
                            c <- c + 1
                            if c = 20 then
                                printfn "%s" "\n"
                                raise Break
                        if (unbox<string> kind) = " " then
                            n <- n + 1
                        else
                            if (unbox<string> kind) = " odd " then
                                n <- n + 2
                            else
                                try
                                    while true do
                                        n <- n + 2
                                        if unbox<bool> (isPrime n) then
                                            raise Break
                                with
                                | Break -> ()
                                | Continue -> ()
                with
                | Break -> ()
                | Continue -> ()
            with
            | Break -> ()
            | Continue -> ()
        let mutable n: int = 7
        let mutable c: int = 0
        while c < 100000 do
            if unbox<bool> (isBrazilian n) then
                c <- c + 1
            n <- n + 1
        printfn "%s" ("The 100,000th Brazilian number: " + (string (n - 1)))
        __ret
    with
        | Return -> __ret
main()
